<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Dave T. Gerrard" />

<meta name="date" content="2017-02-28" />

<title>An introduction to the GenomicLayers package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">An introduction to the GenomicLayers package</h1>
<h4 class="author"><em>Dave T. Gerrard</em></h4>
<h4 class="date"><em>2017-02-28</em></h4>



<p>This will be an introductory vignette on how and why to use GenomicLayers.</p>
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>Rationale…</p>
<p><em>GenomicLayers</em> simulates epi-genetic changes to DNA sequence or entire genomes. It does this by linking the DNA sequence, which cannot be changed, with one or more equally sized <em>layers</em>, which can be modified by <em>binding factors</em>. The <em>binding factors</em> may recognise specific DNA sequences or motifs or they may recognise existing modifications already applied to the layers. After binding, the <em>binding factors</em> may then make modifications to any or all of the layers. In this way, some regions of the sequence or genomes may change states (e.g. open vs closed) and permit or deny access to further binding factors.</p>
</div>
<div id="installation" class="section level3">
<h3>Installation</h3>
<p>You may have already done this but if not, you can install the <em>GenomicLayers</em> package direct from GitHub if you already have the <em>devtools</em> package installed and working. For <em>devtools</em> see here: …</p>
<p>If you already have <em>devtools</em>, then the following commands should install <em>GenomicLayers</em>:-</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">library</span>(devtools)
    <span class="kw">install_github</span>(<span class="dt">repo=</span><span class="st">&quot;davetgerrard/GenomicLayers&quot;</span>,<span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>The above may take several minutes and requires several dependencies. If it does not work, or you are in a hurry, leave out the ‘build_vignettes’. Hint: make sure <em>devtools</em> is up to the most recent version.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">install_github</span>(<span class="dt">repo=</span><span class="st">&quot;davetgerrard/GenomicLayers&quot;</span>)</code></pre></div>
<p>To view the introduction vignette, type</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">vignette</span>(<span class="st">&quot;GenomicLayersIntroduction&quot;</span>)</code></pre></div>
<p>To view what vignettes are available, type</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">vignette</span>(<span class="dt">package=</span><span class="st">&quot;GenomicLayers&quot;</span>)</code></pre></div>
</div>
<div id="set-up-the-target-sequence-or-genome" class="section level3">
<h3>Set up the target sequence or genome</h3>
<p>Load the Biostrings library.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Biostrings)    <span class="co"># (this command outputs a lot of information, which is hidden in this vignette)</span></code></pre></div>
<p>Our sequence can be based on a single DNA sequence or on a entire genome. For individual sequences we convert them to DNAString objects. If the latter, it must be loaded as a BSgenome object. See …</p>
<p><em>GenomicLayers</em> is designed to work on whole eukaryotic genomes including human and mouse but as these are rather large, for the purpose of demonstration we will use a shorter sequence as an example. The <em>Saccharomyces cerevisiae</em> (Yeast) chromosome 1 is included in the <em>Biostrings</em> package, which should have been installed along with <em>GenomicLayers</em></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;yeastSEQCHR1&quot;</span>)   <span class="co"># from Biostrings package</span>
<span class="kw">class</span>(yeastSEQCHR1)</code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nchar</span>(yeastSEQCHR1)</code></pre></div>
<pre><code>## [1] 230208</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">targetSeq &lt;-<span class="st"> </span><span class="kw">DNAString</span>(yeastSEQCHR1)
targetSeq </code></pre></div>
<pre><code>##   230208-letter &quot;DNAString&quot; instance
## seq: CCACACCACACCCACACACCCACACACCACACCA...GTGTGGGTGTGGTGTGGGTGTGGTGTGTGTGGG</code></pre>
<p>The chromosome is now stored as a <em>DNAString</em> object, which gives us access to a range of pattern searching functions provided by the <em>Biostrings</em> package.</p>
<p>To use larger sequences and genomes comprised of multiple chromosomes we expect users to make use of the Bioconductor BSgenome packages. For more on this, see the section <strong>Using larger genomes</strong> below.</p>
</div>
<div id="create-some-layers-on-our-sequence." class="section level3">
<h3>Create some layers on our sequence.</h3>
<p>This is where we use the sequence and build some layers on it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicLayers)
layers1 &lt;-<span class="st"> </span><span class="kw">createLayerList.DNAstring</span>(targetSeq, <span class="dt">layerNames=</span><span class="kw">c</span>(<span class="st">&quot;recruiter&quot;</span>,<span class="st">&quot;promoter&quot;</span>))
layers1</code></pre></div>
<pre><code>## $layerSet
## $layerSet$LAYER.0
##   230208-letter &quot;DNAString&quot; instance
## seq: CCACACCACACCCACACACCCACACACCACACCA...GTGTGGGTGTGGTGTGGGTGTGGTGTGTGTGGG
## 
## $layerSet$recruiter
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## $layerSet$promoter
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## 
## $history
## NULL</code></pre>
</div>
<div id="create-some-binding-factors" class="section level3">
<h3>Create some binding factors</h3>
<p>This is where we specify some binding factors that can bind to the sequence and, optionally, modify the layers (but not the sequence).</p>
<p>If you want to learn about pattern matching against DNA sequences, we recommend that you check out the vignettes in the Biostrings package. Perhaps also have a look at PWMenrich, MOTIV etc.</p>
<p>We first create a binding factor that can recognise a DNA motif for a TATA box, we’ll call it tata-box and then test it against the layerSet object we’ve already created. It should create an IRanges or hits object with the locations for all matches to the sequence TATAA. We’ll also specify that this binding factor should mark those positions on the layer <em>recruiter</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bf1 &lt;-<span class="st"> </span><span class="kw">createBindingFactor.DNA_motif</span>(<span class="dt">name=</span><span class="st">&quot;tata-box&quot;</span>, <span class="dt">patternString =</span> <span class="st">&quot;TATAWAWA&quot;</span>, <span class="dt">profile.layers=</span><span class="ot">NULL</span>, <span class="dt">mod.layers=</span><span class="st">&quot;recruiter&quot;</span>,<span class="dt">mod.marks=</span><span class="dv">1</span>) 
<span class="co">#print(bf1)</span>
<span class="kw">matchBindingFactor</span>(<span class="dt">layerSet=</span>layers1$layerSet, <span class="dt">bindingFactor=</span>bf1)</code></pre></div>
<pre><code>## IRanges object with 63 ranges and 0 metadata columns:
##            start       end     width
##        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      5012      5019         8
##    [2]      5522      5529         8
##    [3]      5975      5982         8
##    [4]      7087      7094         8
##    [5]      7134      7143        10
##    ...       ...       ...       ...
##   [59]    216239    216246         8
##   [60]    218254    218261         8
##   [61]    220878    220887        10
##   [62]    222989    222996         8
##   [63]    227426    227433         8</code></pre>
<p>Then create a second binding factor that does not recognise a particular sequence, but which can bind to marks on the <em>recruiter</em> layer . If we test this binding factor alone, it should produce a hits object with no hits because the <em>recruiter</em> layer starts with not marked regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bf2 &lt;-<span class="st"> </span><span class="kw">createBindingFactor.layer_region</span>(<span class="dt">name=</span><span class="st">&quot;tata-box&quot;</span>, <span class="dt">profile.layers=</span><span class="st">&quot;recruiter&quot;</span>,<span class="dt">profile.marks=</span><span class="dv">1</span>,                                                                        <span class="dt">mod.layers=</span><span class="st">&quot;promoter&quot;</span>,<span class="dt">mod.marks=</span><span class="dv">1</span>,  <span class="dt">patternLength =</span> <span class="dv">5</span>) 
<span class="co">#print(bf2)</span>
<span class="kw">matchBindingFactor</span>(<span class="dt">layerSet=</span>layers1$layerSet, <span class="dt">bindingFactor=</span>bf2)</code></pre></div>
<pre><code>## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</code></pre>
<p>To use multiple binding factors together in a sequence, we add them into a named list. IMPORTANT: currently, this list must have names, so typically we specify the names when building the list. The list names do not have to agree with the <em>$name</em> property of the individual binding factors (this allows the same binding factor to be re-used without re-specifying the model).</p>
<p>(Also, nicer formatting for binding factors when printing)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bf_list &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">bf1=</span>bf1, <span class="dt">bf2=</span>bf2)
<span class="co">#print.bfSet(list(bf1))  # not working in package. FIX this .</span></code></pre></div>
</div>
<div id="run-a-sequence-of-layer-binding" class="section level3">
<h3>Run a sequence of layer binding</h3>
<p>Run a simulation and collect the output</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">modLayers &lt;-<span class="st"> </span><span class="kw">runLayerBinding</span>(<span class="dt">layerList=</span>layers1, <span class="dt">factorSet=</span>bf_list, <span class="dt">verbose=</span><span class="ot">TRUE</span>, <span class="dt">iter=</span><span class="dv">1000</span>)</code></pre></div>
<pre><code>## [1] &quot;2017-02-28 16:05:10 runLayerBinding pos 1&quot;
## [1] &quot;Sequence of length  230208 , using  1 windows of length 230208&quot;
## [1] &quot;bf1 63 63&quot;
## [1] &quot;bf2 63 63&quot;
## [1] &quot;2017-02-28 16:05:10 runLayerBinding.fast pos 2&quot;</code></pre>
<p>Hopefully, the layer called ‘promoter’ should now be marked in some places.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">modLayers$layerSet[[<span class="st">'promoter'</span>]]</code></pre></div>
<pre><code>## IRanges object with 63 ranges and 0 metadata columns:
##            start       end     width
##        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      5013      5017         5
##    [2]      5523      5527         5
##    [3]      5976      5980         5
##    [4]      7088      7092         5
##    [5]      7136      7140         5
##    ...       ...       ...       ...
##   [59]    216240    216244         5
##   [60]    218255    218259         5
##   [61]    220880    220884         5
##   [62]    222990    222994         5
##   [63]    227427    227431         5</code></pre>
</div>
<div id="visualise-the-results" class="section level3">
<h3>Visualise the results</h3>
<p>What happened in the simulation?</p>
<p>Are the marked regions anywhere near the real Yeast promoters? - No.</p>
</div>
<div id="using-larger-genomes" class="section level3">
<h3>Using larger genomes</h3>
<p>Get to know the BSgenome packages….</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
