<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Dave T. Gerrard" />

<meta name="date" content="2023-08-29" />

<title>An introduction to the GenomicLayers package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">An introduction to the GenomicLayers
package</h1>
<h4 class="author">Dave T. Gerrard</h4>
<h4 class="date">2023-08-29</h4>



<p>This will be an introductory vignette on how and why to use
GenomicLayers.</p>
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>Rationale…</p>
<p><em>GenomicLayers</em> simulates epi-genetic changes to DNA sequence
or entire genomes. It does this by linking the DNA sequence, which
cannot be changed, with one or more equally sized <em>layers</em>, which
can be modified by <em>binding factors</em>. The <em>binding
factors</em> may recognise specific DNA sequences or motifs or they may
recognise existing modifications already applied to the layers. After
binding, the <em>binding factors</em> may then make modifications to any
or all of the layers. In this way, some regions of the sequence or
genomes may change states (e.g. open vs closed) and permit or deny
access to further binding factors.</p>
</div>
<div id="installation" class="section level3">
<h3>Installation</h3>
<p>You may have already done this but if not, you can install the
<em>GenomicLayers</em> package direct from GitHub if you already have
the <em>devtools</em> package installed and working. For
<em>devtools</em> see here: …</p>
<p>If you already have <em>devtools</em>, then the following commands
should install <em>GenomicLayers</em>:-</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(devtools)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install_github</span>(<span class="at">repo=</span><span class="st">&quot;davetgerrard/GenomicLayers&quot;</span>,<span class="at">build_vignettes =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The above may take several minutes and requires several dependencies.
If it does not work, or you are in a hurry, leave out the
‘build_vignettes’. Hint: make sure <em>devtools</em> is up to the most
recent version.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install_github</span>(<span class="at">repo=</span><span class="st">&quot;davetgerrard/GenomicLayers&quot;</span>)</span></code></pre></div>
<p>To view the introduction vignette, type</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">vignette</span>(<span class="st">&quot;GenomicLayersIntroduction&quot;</span>)</span></code></pre></div>
<p>To view what vignettes are available, type</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">vignette</span>(<span class="at">package=</span><span class="st">&quot;GenomicLayers&quot;</span>)</span></code></pre></div>
</div>
<div id="set-up-the-target-sequence-or-genome" class="section level3">
<h3>Set up the target sequence or genome</h3>
<p>Load the Biostrings library.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)    <span class="co"># (this command outputs a lot of information, which is hidden in this vignette)</span></span></code></pre></div>
<p>Our sequence can be based on a single DNA sequence or on a entire
genome. For individual sequences we convert them to DNAString objects.
If the latter, it must be loaded as a BSgenome object. See …</p>
<p><em>GenomicLayers</em> is designed to work on whole eukaryotic
genomes including human and mouse but as these are rather large, for the
purpose of demonstration we will use a shorter sequence as an example.
The <em>Saccharomyces cerevisiae</em> (Yeast) chromosome 1 is included
in the <em>Biostrings</em> package, which should have been installed
along with <em>GenomicLayers</em></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;yeastSEQCHR1&quot;</span>)   <span class="co"># from Biostrings package</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(yeastSEQCHR1)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(yeastSEQCHR1)</span></code></pre></div>
<pre><code>## [1] 230208</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>targetSeq <span class="ot">&lt;-</span> <span class="fu">DNAString</span>(yeastSEQCHR1)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>targetSeq </span></code></pre></div>
<pre><code>## 230208-letter DNAString object
## seq: CCACACCACACCCACACACCCACACACCACACCACA...GGTGTGTGGGTGTGGTGTGGGTGTGGTGTGTGTGGG</code></pre>
<p>The chromosome is now stored as a <em>DNAString</em> object, which
gives us access to a range of pattern searching functions provided by
the <em>Biostrings</em> package.</p>
<p>To use larger sequences and genomes comprised of multiple chromosomes
we expect users to make use of the Bioconductor BSgenome packages. For
more on this, see the section <strong>Using larger genomes</strong>
below.</p>
</div>
<div id="create-some-layers-on-our-sequence." class="section level3">
<h3>Create some layers on our sequence.</h3>
<p>This is where we use the sequence and build some layers on it.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicLayers)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>layers1 <span class="ot">&lt;-</span> <span class="fu">createLayerList.DNAstring</span>(targetSeq, <span class="at">layerNames=</span><span class="fu">c</span>(<span class="st">&quot;recruiter&quot;</span>,<span class="st">&quot;promoter&quot;</span>))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>layers1</span></code></pre></div>
<pre><code>## $layerSet
## $layerSet$LAYER.0
## 230208-letter DNAString object
## seq: CCACACCACACCCACACACCCACACACCACACCACA...GGTGTGTGGGTGTGGTGTGGGTGTGGTGTGTGTGGG
## 
## $layerSet$recruiter
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## $layerSet$promoter
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## 
## $history
## NULL</code></pre>
</div>
<div id="create-some-binding-factors" class="section level3">
<h3>Create some binding factors</h3>
<p>This is where we specify some binding factors that can bind to the
sequence and, optionally, modify the layers (but not the sequence).</p>
<p>If you want to learn about pattern matching against DNA sequences, we
recommend that you check out the vignettes in the Biostrings package.
Perhaps also have a look at PWMenrich, MOTIV etc.</p>
<p>We first create a binding factor that can recognise a DNA motif for a
TATA box, we’ll call it tata-box and then test it against the layerSet
object we’ve already created. It should create an IRanges or hits object
with the locations for all matches to the sequence TATAA. We’ll also
specify that this binding factor should mark those positions on the
layer <em>recruiter</em>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>bf1 <span class="ot">&lt;-</span> <span class="fu">createBindingFactor.DNA_consensus</span>(<span class="at">name=</span><span class="st">&quot;tata-box&quot;</span>, <span class="at">patternString =</span> <span class="st">&quot;TATAWAWA&quot;</span>, <span class="at">profile.layers=</span><span class="cn">NULL</span>, <span class="at">mod.layers=</span><span class="st">&quot;recruiter&quot;</span>,<span class="at">mod.marks=</span><span class="dv">1</span>) </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print(bf1)</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matchBindingFactor</span>(<span class="at">layerSet=</span>layers1<span class="sc">$</span>layerSet, <span class="at">bindingFactor=</span>bf1)</span></code></pre></div>
<pre><code>## IRanges object with 63 ranges and 0 metadata columns:
##            start       end     width
##        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      5012      5019         8
##    [2]      5522      5529         8
##    [3]      5975      5982         8
##    [4]      7087      7094         8
##    [5]      7134      7143        10
##    ...       ...       ...       ...
##   [59]    216239    216246         8
##   [60]    218254    218261         8
##   [61]    220878    220887        10
##   [62]    222989    222996         8
##   [63]    227426    227433         8</code></pre>
<p>Then create a second binding factor that does not recognise a
particular sequence, but which can bind to marks on the
<em>recruiter</em> layer . If we test this binding factor alone, it
should produce a hits object with no hits because the <em>recruiter</em>
layer starts with not marked regions.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>bf2 <span class="ot">&lt;-</span> <span class="fu">createBindingFactor.layer_region</span>(<span class="at">name=</span><span class="st">&quot;tata-box&quot;</span>, <span class="at">profile.layers=</span><span class="st">&quot;recruiter&quot;</span>,<span class="at">profile.marks=</span><span class="dv">1</span>,                                                                        <span class="at">mod.layers=</span><span class="st">&quot;promoter&quot;</span>,<span class="at">mod.marks=</span><span class="dv">1</span>,  <span class="at">patternLength =</span> <span class="dv">5</span>) </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print(bf2)</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matchBindingFactor</span>(<span class="at">layerSet=</span>layers1<span class="sc">$</span>layerSet, <span class="at">bindingFactor=</span>bf2)</span></code></pre></div>
<pre><code>## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</code></pre>
<p>To use multiple binding factors together in a sequence, we add them
into a named list. IMPORTANT: currently, this list must have names, so
typically we specify the names when building the list. The list names do
not have to agree with the <em>$name</em> property of the individual
binding factors (this allows the same binding factor to be re-used
without re-specifying the model).</p>
<p>(Also, nicer formatting for binding factors when printing)</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>bf_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">bf1=</span>bf1, <span class="at">bf2=</span>bf2)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#print.bfSet(list(bf1))  # not working in package. FIX this .</span></span></code></pre></div>
</div>
<div id="run-a-sequence-of-layer-binding" class="section level3">
<h3>Run a sequence of layer binding</h3>
<p>Run a simulation and collect the output</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>modLayers <span class="ot">&lt;-</span> <span class="fu">runLayerBinding</span>(<span class="at">layerList=</span>layers1, <span class="at">factorSet=</span>bf_list, <span class="at">verbose=</span><span class="cn">TRUE</span>, <span class="at">iter=</span><span class="dv">1000</span>)</span></code></pre></div>
<pre><code>## [1] &quot;2023-08-29 22:22:54 runLayerBinding pos 1&quot;
## [1] &quot;Sequence of length  230208 , using  1 windows of length 230208&quot;
## [1] &quot;bf1 63 63&quot;
## [1] &quot;bf2 63 63&quot;
## [1] &quot;2023-08-29 22:22:54 runLayerBinding.fast pos 2&quot;</code></pre>
<p>Hopefully, the layer called ‘promoter’ should now be marked in some
places.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>modLayers<span class="sc">$</span>layerSet[[<span class="st">&#39;promoter&#39;</span>]]</span></code></pre></div>
<pre><code>## IRanges object with 63 ranges and 0 metadata columns:
##            start       end     width
##        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      5013      5017         5
##    [2]      5523      5527         5
##    [3]      5976      5980         5
##    [4]      7088      7092         5
##    [5]      7136      7140         5
##    ...       ...       ...       ...
##   [59]    216240    216244         5
##   [60]    218255    218259         5
##   [61]    220880    220884         5
##   [62]    222990    222994         5
##   [63]    227427    227431         5</code></pre>
</div>
<div id="visualise-the-results" class="section level3">
<h3>Visualise the results</h3>
<p>What happened in the simulation?</p>
<p>Are the marked regions anywhere near the real Yeast promoters? -
No.</p>
</div>
<div id="using-larger-genomes" class="section level3">
<h3>Using larger genomes</h3>
<p>Get to know the BSgenome packages….</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
