<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Dave T. Gerrard" />

<meta name="date" content="2023-08-29" />

<title>Yeast RIP methylation driven mutation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Yeast RIP methylation driven mutation</h1>
<h4 class="author">Dave T. Gerrard</h4>
<h4 class="date">2023-08-29</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Some fungal species silence transposable elements in their genome by
recognising the high local AT sequence content of the repeats and then
targetting DNA methylation to Cytosines within these regions. Some
species contain a RID mechanism to then “mutate” methylated cytosines
and permanently silence the transposons. See <span class="citation">(He
et al. 2020)</span> In this vignette, we will simulate the two-step
recruitment of silencing machinery and methylation but not the mutation.
We will use the S.cerevisae genome even though it does not feature this
silencing mechanism but because the genome is small and available as a
BSgenome package.</p>
<p>This is a citation<span class="citation">(He et al. 2020)</span> and
this is a (link to the paper)[ <a href="https://www.mdpi.com/2076-2607/8/2/227" class="uri">https://www.mdpi.com/2076-2607/8/2/227</a>]</p>
<p>First we load required libraries. The output is hidden in this
vignette as you may see many messages caused by loading dependent
packages.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicLayers)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BSgenome.Scerevisiae.UCSC.sacCer3)</span></code></pre></div>
<p>Create a named link to the genome, inspect it and remind ourselves of
the degenerate IUPAC codes.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>genome <span class="ot">&lt;-</span> BSgenome.Scerevisiae.UCSC.sacCer3</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">seqinfo</span>(genome)</span></code></pre></div>
<pre><code>## Seqinfo object with 17 sequences (1 circular) from sacCer3 genome:
##   seqnames seqlengths isCircular  genome
##   chrI         230218      FALSE sacCer3
##   chrII        813184      FALSE sacCer3
##   chrIII       316620      FALSE sacCer3
##   chrIV       1531933      FALSE sacCer3
##   chrV         576874      FALSE sacCer3
##   ...             ...        ...     ...
##   chrXIII      924431      FALSE sacCer3
##   chrXIV       784333      FALSE sacCer3
##   chrXV       1091291      FALSE sacCer3
##   chrXVI       948066      FALSE sacCer3
##   chrM          85779       TRUE sacCer3</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>IUPAC_CODE_MAP   <span class="co"># check the IUPAC codes.</span></span></code></pre></div>
<pre><code>##      A      C      G      T      M      R      W      S      Y      K      V 
##    &quot;A&quot;    &quot;C&quot;    &quot;G&quot;    &quot;T&quot;   &quot;AC&quot;   &quot;AG&quot;   &quot;AT&quot;   &quot;CG&quot;   &quot;CT&quot;   &quot;GT&quot;  &quot;ACG&quot; 
##      H      D      B      N 
##  &quot;ACT&quot;  &quot;AGT&quot;  &quot;CGT&quot; &quot;ACGT&quot;</code></pre>
</div>
<div id="create-a-layerset-object-to-store-results-linked-to-the-genome." class="section level2">
<h2>Create a “layerSet” object to store results linked to the
genome.</h2>
<p>We will want to store hits on the genome as Layers that can act as
context for the selection of further sites of activity. These could be
areas of open or closed chromatin, specific epigenetic marks, or any
user-defined region of activity. Here we will create layers to represent
the presence of H3K9me3 and DNA-methylation. In this simulation, both
layers begin with no regions, but they could be pre-populated with
GRanges objects.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sacCer.LS <span class="ot">&lt;-</span> <span class="fu">createLayerSet.BSgenome</span>(BSgenome.Scerevisiae.UCSC.sacCer3,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">n.layer=</span><span class="dv">2</span>, </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">layer.names=</span><span class="fu">c</span>(<span class="st">&quot;H3K9me3&quot;</span>, <span class="st">&quot;DNA-meth&quot;</span>))</span></code></pre></div>
</div>
<div id="create-binding-factors" class="section level2">
<h2>Create binding factors</h2>
<p>The first binding factor will bind to regions of the genome of 400bp
that have an AT% greater than or equal to 70%. Where it binds, it will
set the H3K9me3 layer to state 1 (rather than 0 or empty).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>DIM5.ATpc <span class="ot">&lt;-</span> <span class="fu">createBindingFactor.DNA_consensus</span>(<span class="at">name=</span><span class="st">&quot;DIM5.ATpc&quot;</span>,    </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">patternString =</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(<span class="st">&quot;W&quot;</span>, <span class="dv">400</span>), <span class="at">collapse =</span><span class="st">&quot;&quot;</span>), </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fixed=</span><span class="cn">FALSE</span> , <span class="at">max.mismatch=</span><span class="dv">400</span><span class="sc">*</span>.<span class="dv">3</span>  ,  </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">profile.layers=</span><span class="fu">c</span>(<span class="st">&quot;H3K9me3&quot;</span>, <span class="st">&quot;DNA-meth&quot;</span>),</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">profile.marks=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">mod.layers=</span><span class="fu">c</span>(<span class="st">&quot;H3K9me3&quot;</span>), <span class="at">mod.marks=</span><span class="dv">1</span> )</span></code></pre></div>
<p>It is worth checking what a binding factor will bind and we can use
the function matchBindingFactor.BSgenome for this.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check DIM5.ATpc is finding some regions.</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>(testtemp <span class="ot">&lt;-</span> <span class="fu">matchBindingFactor.BSgenome</span>( <span class="at">layerSet=</span>sacCer.LS, </span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">bindingFactor=</span>DIM5.ATpc))</span></code></pre></div>
<pre><code>## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.</code></pre>
<pre><code>## GRanges object with 730332 ranges and 0 metadata columns:
##            seqnames      ranges strand
##               &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##        [1]     chrI   4327-4726      *
##        [2]     chrI   4328-4727      *
##        [3]     chrI   4329-4728      *
##        [4]     chrI   4330-4729      *
##        [5]     chrI   4331-4730      *
##        ...      ...         ...    ...
##   [730328]     chrM 85376-85775      *
##   [730329]     chrM 85377-85776      *
##   [730330]     chrM 85378-85777      *
##   [730331]     chrM 85379-85778      *
##   [730332]     chrM 85380-85779      *
##   -------
##   seqinfo: 17 sequences (1 circular) from sacCer3 genome</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check that they have high AT richness</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getSeq</span>(genome, <span class="at">names=</span>testtemp[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>## DNAStringSet object of length 5:
##     width seq
## [1]   400 ACTACATGACAAGCAACTCATAATTTAAGTGGAT...TGGATAACCAGATCGTACATCTCTGATAAACAA
## [2]   400 CTACATGACAAGCAACTCATAATTTAAGTGGATA...GGATAACCAGATCGTACATCTCTGATAAACAAT
## [3]   400 TACATGACAAGCAACTCATAATTTAAGTGGATAG...GATAACCAGATCGTACATCTCTGATAAACAATC
## [4]   400 ACATGACAAGCAACTCATAATTTAAGTGGATAGC...ATAACCAGATCGTACATCTCTGATAAACAATCT
## [5]   400 CATGACAAGCAACTCATAATTTAAGTGGATAGCT...TAACCAGATCGTACATCTCTGATAAACAATCTT</code></pre>
<p>Create a second binding factor to recognise regions that can be
methylated (and subsequently mutated). RIP mutation primarily happens at
(A/Tp)Cp(A/T) so we will use the degenerate code “WCW” where “W” can
mean “A” or “T”. Combine both binding factors into a list ready to use
them together.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>RID <span class="ot">&lt;-</span> <span class="fu">createBindingFactor.DNA_consensus</span>(<span class="at">name=</span><span class="st">&quot;RID&quot;</span>,    </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">patternString =</span> <span class="st">&quot;WCW&quot;</span>, <span class="at">fixed=</span><span class="cn">FALSE</span> ,   </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">max.mismatch=</span><span class="dv">0</span>  ,  <span class="at">profile.layers=</span><span class="fu">c</span>(<span class="st">&quot;H3K9me3&quot;</span>),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">profile.marks=</span><span class="dv">1</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">mod.layers=</span><span class="fu">c</span>(<span class="st">&quot;DNA-meth&quot;</span>), <span class="at">mod.marks=</span><span class="dv">1</span> , <span class="at">stateWidth=</span><span class="dv">1</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>                                  </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>bfList <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">DIM5.ATpc=</span>DIM5.ATpc, <span class="at">RID=</span>RID)                                  </span></code></pre></div>
</div>
<div id="the-simulation" class="section level2">
<h2>The simulation</h2>
<p>Now we have the layerSet and the list of binding factors, we are
ready to run the simulation. The default behaviour is for each binding
factor to be applied in turn (in the order of the list) and for 1000
suitable regions to be selected for each.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>resultLayerGenome <span class="ot">&lt;-</span> <span class="fu">runLayerBinding.BSgenome</span>(  <span class="at">layerList=</span>sacCer.LS, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">factorSet=</span> bfList)</span></code></pre></div>
<pre><code>## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.

## Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 232 out-of-bound ranges located on sequence
##   chrM. Note that ranges located on a sequence whose length is unknown
##   (NA) or on a circular sequence are not considered out-of-bound (use
##   seqlengths() and isCircular() to get the lengths and circularity flags
##   of the underlying sequences). You can use trim() to trim these ranges.
##   See ?`trim,GenomicRanges-method` for more information.</code></pre>
<pre><code>## ..[1] &quot;&quot;</code></pre>
<p>The object resultLayerGenome is a list and now contains marks on the
layers H3K9me3 and DNA-meth. Each set of results is GRanges object and
can be extracted and compared using the many functions available in the
GenomicRanges package.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>resultLayerGenome</span></code></pre></div>
<pre><code>## $layerSet
## $layerSet$LAYER.0
## Yeast genome:
## # organism: Saccharomyces cerevisiae (Yeast)
## # genome: sacCer3
## # provider: UCSC
## # release date: April 2011
## # 17 sequences:
## #   chrI    chrII   chrIII  chrIV   chrV    chrVI   chrVII  chrVIII chrIX  
## #   chrX    chrXI   chrXII  chrXIII chrXIV  chrXV   chrXVI  chrM           
## # (use &#39;seqnames()&#39; to see all the sequence names, use the &#39;$&#39; or &#39;[[&#39; operator
## # to access a given sequence)
## 
## $layerSet$H3K9me3
## GRanges object with 1427 ranges and 0 metadata columns:
##          seqnames        ranges strand
##             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;
##      [1]     chrI     4406-4805      *
##      [2]     chrI     6466-7245      *
##      [3]     chrI    9857-10491      *
##      [4]     chrI   10963-11612      *
##      [5]     chrI   16659-17598      *
##      ...      ...           ...    ...
##   [1423]   chrXVI 937189-937606      *
##   [1424]   chrXVI 942101-942594      *
##   [1425]     chrM      45-14110      *
##   [1426]     chrM   14161-16579      *
##   [1427]     chrM   16787-85757      *
##   -------
##   seqinfo: 17 sequences (1 circular) from sacCer3 genome
## 
## $layerSet$`DNA-meth`
## GRanges object with 10000 ranges and 0 metadata columns:
##           seqnames    ranges strand
##              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##       [1]     chrI      4553      *
##       [2]     chrI      4605      *
##       [3]     chrI      4658      *
##       [4]     chrI      4728      *
##       [5]     chrI      6565      *
##       ...      ...       ...    ...
##    [9996]     chrM     83358      *
##    [9997]     chrM     83953      *
##    [9998]     chrM     84696      *
##    [9999]     chrM     85058      *
##   [10000]     chrM     85397      *
##   -------
##   seqinfo: 17 sequences (1 circular) from sacCer3 genome
## 
## 
## $cache
## $cache$DIM5.ATpc
## $cache$DIM5.ATpc$LAYER.0
## GRanges object with 730332 ranges and 0 metadata columns:
##            seqnames      ranges strand
##               &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##        [1]     chrI   4327-4726      *
##        [2]     chrI   4328-4727      *
##        [3]     chrI   4329-4728      *
##        [4]     chrI   4330-4729      *
##        [5]     chrI   4331-4730      *
##        ...      ...         ...    ...
##   [730328]     chrM 85376-85775      *
##   [730329]     chrM 85377-85776      *
##   [730330]     chrM 85378-85777      *
##   [730331]     chrM 85379-85778      *
##   [730332]     chrM 85380-85779      *
##   -------
##   seqinfo: 17 sequences (1 circular) from sacCer3 genome
## 
## 
## $cache$RID
## $cache$RID$LAYER.0
## GRanges object with 1764582 ranges and 0 metadata columns:
##             seqnames      ranges strand
##                &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##         [1]     chrI         3-5      *
##         [2]     chrI        8-10      *
##         [3]     chrI       14-16      *
##         [4]     chrI       16-18      *
##         [5]     chrI       22-24      *
##         ...      ...         ...    ...
##   [1764578]     chrM 85671-85673      *
##   [1764579]     chrM 85677-85679      *
##   [1764580]     chrM 85705-85707      *
##   [1764581]     chrM 85731-85733      *
##   [1764582]     chrM 85747-85749      *
##   -------
##   seqinfo: 17 sequences (1 circular) from sacCer3 genome</code></pre>
<p>The sub-list “$cache” contains a record of possible matches for
binding factors that recognise LAYER.0 (the genome). Further simulations
using the same result object (resultLayerGenome) can re-use this cache
and will save some time.</p>
<p>This vignette contains just two parts of this repressive
mechanism.</p>
<ul>
<li>H3K9me3 at areas of high AT DIM-5/KMT1</li>
<li>Methylation of Cs within these regions. DIM-2</li>
</ul>
<p>It could be further extended to incorporate spreading of the H3K9me3
mark by HP1. At time of writing, the BSgenome is immutable, so we cannot
“mutate” the methylated Cytosines.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-he_pattern_2020" class="csl-entry">
He, Chang, Zhanquan Zhang, Boqiang Li, and Shiping Tian. 2020.
<span>“The <span>Pattern</span> and <span>Function</span> of
<span>DNA</span> <span>Methylation</span> in <span>Fungal</span>
<span>Plant</span> <span>Pathogens</span>.”</span>
<em>Microorganisms</em> 8 (2): 227. <a href="https://doi.org/10.3390/microorganisms8020227">https://doi.org/10.3390/microorganisms8020227</a>.
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
